// objection/lib/relations/hasMany/HasManyRelation.js
'use strict';

const { Relation } = require('objection/lib/relations/Relation');
const { HasManyInsertOperation } = require('objection/lib/relations/hasMany/HasManyInsertOperation');
const { HasManyRelateOperation } = require('objection/lib/relations/hasMany/HasManyRelateOperation');
const { HasManyUnrelateOperation } = require('objection/lib/relations/hasMany/HasManyUnrelateOperation');

class HasManyRelation extends Relation {
  insert(_, owner) {
    return new HasManyInsertOperation('insert', {
      relation: this,
      owner: owner,
    });
  }

  relate(_, owner) {
    return new HasManyRelateOperation('relate', {
      relation: this,
      owner: owner,
    });
  }

  unrelate(_, owner) {
    return new HasManyUnrelateOperation('unrelate', {
      relation: this,
      owner: owner,
    });
  }

  hasRelateProp(model) {
    return model.$hasId();
  }

  setRelateProp(model, values) {
    model.$id(values);
  }
}

Object.defineProperties(HasManyRelation.prototype, {
  isObjectionHasManyRelation: {
    enumerable: false,
    writable: false,
    value: true,
  },
});

module.exports = {
  HasManyRelation,
};
