// objection/lib/queryBuilder/operations/InstanceUpdateOperation.js
'use strict';

const { UpdateOperation } = require('objection/lib/queryBuilder/operations/UpdateOperation');
const { InstanceFindOperation } = require('objection/lib/queryBuilder/operations/InstanceFindOperation');
const { assertHasId } = require('objection/lib/utils/assert');
const { isObject } = require('objection/lib/utils/objectUtils');

class InstanceUpdateOperation extends UpdateOperation {
  constructor(name, opt) {
    super(name, opt);

    this.instance = opt.instance;
    this.modelOptions.old = opt.instance;
  }

  onAdd(builder, args) {
    const retVal = super.onAdd(builder, args);

    if (!this.model) {
      this.model = this.instance;
    }

    return retVal;
  }

  onBuild(builder) {
    super.onBuild(builder);

    assertHasId(this.instance);
    builder.findById(this.instance.$id());
  }

  async onAfter2(builder, result) {
    // The result may be an object if `returning` was used.
    if (Array.isArray(result)) {
      result = result[0];
    }

    result = await super.onAfter2(builder, result);
    this.instance.$set(this.model);

    if (isObject(result)) {
      this.instance.$set(result);
    }

    return result;
  }

  toFindOperation() {
    return new InstanceFindOperation('find', {
      instance: this.instance,
    });
  }
}

module.exports = {
  InstanceUpdateOperation,
};
