// @interledger/pay/dist/src/controllers/asset-details.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AssetDetailsController = exports.isValidAssetScale = exports.isValidAssetDetails = void 0;
const packet_1 = require("ilp-protocol-stream/dist/src/packet");
const __1 = require("@interledger/pay");
const isValidAssetDetails = (o) => typeof o === 'object' && o !== null && typeof o.code === 'string' && (0, exports.isValidAssetScale)(o.scale);
exports.isValidAssetDetails = isValidAssetDetails;
const isValidAssetScale = (o) => typeof o === 'number' && o >= 0 && o <= 255 && Number.isInteger(o);
exports.isValidAssetScale = isValidAssetScale;
class AssetDetailsController {
    constructor({ destinationAsset }) {
        this.destinationAsset = destinationAsset;
    }
    getDestinationAsset() {
        return this.destinationAsset;
    }
    applyRequest() {
        return ({ frames, log }) => {
            const newAssetDetails = (frames !== null && frames !== void 0 ? frames : [])
                .filter((frame) => frame.type === packet_1.FrameType.ConnectionAssetDetails)
                .map((frame) => ({
                code: frame.sourceAssetCode,
                scale: frame.sourceAssetScale,
            }));
            for (const { code: assetCode, scale: assetScale } of newAssetDetails) {
                if (!this.destinationAsset) {
                    log.debug('got destination asset details: %s %s', assetCode, assetScale);
                    this.destinationAsset = {
                        code: assetCode,
                        scale: assetScale,
                    };
                }
                else if (this.destinationAsset.code !== assetCode ||
                    this.destinationAsset.scale !== assetScale) {
                    log.error('ending payment: remote unexpectedly changed destination asset from %s %s to %s %s', this.destinationAsset.code, this.destinationAsset.scale, assetCode, assetScale);
                    return __1.PaymentError.DestinationAssetConflict;
                }
            }
        };
    }
}
exports.AssetDetailsController = AssetDetailsController;
