// @interledger/pay/dist/src/controllers/establishment.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EstablishmentController = void 0;
const _1 = require("@interledger/pay/dist/src/controllers");
const packet_1 = require("ilp-protocol-stream/dist/src/packet");
const payment_1 = require("@interledger/pay/dist/src/senders/payment");
class EstablishmentController {
    constructor({ destinationAddress }) {
        this.isConnected = false;
        this.destinationAddress = destinationAddress;
    }
    didConnect() {
        return this.isConnected;
    }
    buildRequest(request) {
        request.setDestinationAddress(this.destinationAddress);
        if (!this.isConnected) {
            request.addFrames(new packet_1.ConnectionMaxStreamIdFrame(payment_1.PaymentSender.DEFAULT_STREAM_ID), new packet_1.ConnectionMaxDataFrame(0));
        }
        return _1.RequestState.Ready();
    }
    applyRequest() {
        return (reply) => {
            this.isConnected = this.isConnected || reply.isAuthentic();
        };
    }
}
exports.EstablishmentController = EstablishmentController;
