// @interledger/pay/dist/src/controllers/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SendState = exports.RequestState = exports.SendStateType = void 0;
var SendStateType;
(function (SendStateType) {
    SendStateType[SendStateType["Done"] = 0] = "Done";
    SendStateType[SendStateType["Error"] = 1] = "Error";
    SendStateType[SendStateType["Schedule"] = 2] = "Schedule";
    SendStateType[SendStateType["Yield"] = 3] = "Yield";
    SendStateType[SendStateType["Ready"] = 4] = "Ready";
    SendStateType[SendStateType["Send"] = 5] = "Send";
})(SendStateType = exports.SendStateType || (exports.SendStateType = {}));
const Error = (error) => ({
    type: SendStateType.Error,
    value: error,
});
const Schedule = (delay) => ({
    type: SendStateType.Schedule,
    delay: delay !== null && delay !== void 0 ? delay : Promise.resolve(),
});
const Yield = () => ({ type: SendStateType.Yield });
const Done = (value) => ({
    type: SendStateType.Done,
    value,
});
const Ready = () => ({
    type: SendStateType.Ready,
});
const Send = (applyReply) => ({
    type: SendStateType.Send,
    applyReply,
});
exports.RequestState = {
    Ready,
    Error,
    Schedule,
    Yield,
};
exports.SendState = {
    Done,
    Error,
    Schedule,
    Yield,
    Send,
};
