// @interledger/pay/dist/src/controllers/sequence.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SequenceController = exports.Counter = void 0;
const _1 = require("@interledger/pay/dist/src/controllers");
const __1 = require("@interledger/pay");
const utils_1 = require("@interledger/pay/dist/src/utils");
class Counter {
    constructor(count) {
        this.count = count;
    }
    static from(count) {
        if ((0, utils_1.isNonNegativeInteger)(count)) {
            return new Counter(count);
        }
    }
    increment() {
        this.count++;
    }
    getCount() {
        return this.count;
    }
}
exports.Counter = Counter;
class SequenceController {
    constructor(counter) {
        this.counter = counter;
    }
    buildRequest(request) {
        if (this.counter.getCount() >= SequenceController.PACKET_LIMIT) {
            request.log.error('ending payment: cannot exceed max safe sequence number.');
            return _1.RequestState.Error(__1.PaymentError.MaxSafeEncryptionLimit);
        }
        else {
            request.setSequence(this.counter.getCount());
            return _1.RequestState.Ready();
        }
    }
    applyRequest() {
        this.counter.increment();
        return;
    }
}
exports.SequenceController = SequenceController;
SequenceController.PACKET_LIMIT = (2 ** 31);
