// @interledger/pay/dist/src/payment-pointer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AccountUrl = exports.createHttpUrl = void 0;
const createHttpUrl = (rawUrl, base) => {
    try {
        const url = new URL(rawUrl, base);
        if (url.protocol === 'https:' || url.protocol === 'http:') {
            return url;
        }
    }
    catch (_) {
        return;
    }
};
exports.createHttpUrl = createHttpUrl;
class AccountUrl {
    constructor(url) {
        this.protocol = url.protocol;
        this.hostname = url.hostname;
        const pathname = url.pathname.replace(/\/$/, '');
        if (!(pathname === '' || pathname === AccountUrl.DEFAULT_PATH)) {
            this.path = pathname;
        }
        this.suffix = url.search + url.hash;
    }
    static fromPaymentPointer(paymentPointer) {
        if (!paymentPointer.startsWith('$')) {
            return;
        }
        const url = (0, exports.createHttpUrl)('https://' + paymentPointer.substring(1));
        if (!url ||
            url.username !== '' ||
            url.password !== '' ||
            url.port !== '' ||
            url.search !== '' ||
            url.hash !== '') {
            return;
        }
        return new AccountUrl(url);
    }
    static fromUrl(rawUrl) {
        const url = (0, exports.createHttpUrl)(rawUrl);
        if (!url || url.username !== '' || url.password !== '' || url.port !== '') {
            return;
        }
        return new AccountUrl(url);
    }
    toEndpointUrl() {
        var _a;
        return (this.protocol + '//' + this.hostname + ((_a = this.path) !== null && _a !== void 0 ? _a : AccountUrl.DEFAULT_PATH) + this.suffix);
    }
    toBaseUrl() {
        var _a;
        return this.protocol + '//' + this.hostname + ((_a = this.path) !== null && _a !== void 0 ? _a : AccountUrl.DEFAULT_PATH);
    }
    toString() {
        return this.toEndpointUrl();
    }
    toPaymentPointer() {
        var _a;
        if (this.protocol !== 'https:' || this.suffix !== '')
            return;
        return '$' + this.hostname + ((_a = this.path) !== null && _a !== void 0 ? _a : '');
    }
}
exports.AccountUrl = AccountUrl;
AccountUrl.DEFAULT_PATH = '/.well-known/pay';
