// objection/lib/relations/belongsToOne/BelongsToOneRelation.js
'use strict';

const { Relation } = require('objection/lib/relations/Relation');
const { BelongsToOneInsertOperation } = require('objection/lib/relations/belongsToOne/BelongsToOneInsertOperation');
const { BelongsToOneDeleteOperation } = require('objection/lib/relations/belongsToOne/BelongsToOneDeleteOperation');
const { BelongsToOneRelateOperation } = require('objection/lib/relations/belongsToOne/BelongsToOneRelateOperation');
const { BelongsToOneUnrelateOperation } = require('objection/lib/relations/belongsToOne/BelongsToOneUnrelateOperation');

class BelongsToOneRelation extends Relation {
  isOneToOne() {
    return true;
  }

  insert(_, owner) {
    return new BelongsToOneInsertOperation('insert', {
      relation: this,
      owner,
    });
  }

  delete(_, owner) {
    return new BelongsToOneDeleteOperation('delete', {
      relation: this,
      owner,
    });
  }

  relate(_, owner) {
    return new BelongsToOneRelateOperation('relate', {
      relation: this,
      owner,
    });
  }

  unrelate(_, owner) {
    return new BelongsToOneUnrelateOperation('unrelate', {
      relation: this,
      owner,
    });
  }
}

Object.defineProperties(BelongsToOneRelation.prototype, {
  isObjectionBelongsToOneRelation: {
    enumerable: false,
    writable: false,
    value: true,
  },
});

module.exports = {
  BelongsToOneRelation,
};
