// @interledger/pay/dist/src/senders/asset-probe.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AssetProbe = void 0;
const controllers_1 = require("@interledger/pay/dist/src/controllers");
const asset_details_1 = require("@interledger/pay/dist/src/controllers/asset-details");
const __1 = require("@interledger/pay");
const packet_1 = require("ilp-protocol-stream/dist/src/packet");
const establishment_1 = require("@interledger/pay/dist/src/controllers/establishment");
const expiry_1 = require("@interledger/pay/dist/src/controllers/expiry");
const sequence_1 = require("@interledger/pay/dist/src/controllers/sequence");
const _1 = require("@interledger/pay/dist/src/senders");
class AssetProbe extends _1.StreamSender {
    constructor(plugin, destination, counter) {
        super(plugin, destination);
        this.requestCount = 0;
        this.replyCount = 0;
        this.establishmentController = new establishment_1.EstablishmentController(destination);
        this.assetController = new asset_details_1.AssetDetailsController(destination);
        this.controllers = [
            new sequence_1.SequenceController(counter),
            this.establishmentController,
            new expiry_1.ExpiryController(),
            this.assetController,
        ];
    }
    nextState(request) {
        const assetDetails = this.assetController.getDestinationAsset();
        if (assetDetails) {
            return controllers_1.SendState.Done(assetDetails);
        }
        if (this.requestCount === 0) {
            request.addFrames(new packet_1.ConnectionNewAddressFrame('')).build();
            request.log.debug('requesting asset details (1 of 2).');
        }
        else if (this.requestCount === 1) {
            const segments = request.destinationAddress.split('.');
            const destinationAddress = [...segments.slice(0, -1), '_', ...segments.slice(-1)]
                .join('.')
                .substring(0, 1023);
            request
                .addFrames(new packet_1.ConnectionNewAddressFrame('private.SEND_ONLY_CLIENT'))
                .setDestinationAddress(destinationAddress)
                .build();
            request.log.debug('requesting asset details (2 of 2).');
        }
        else {
            return controllers_1.SendState.Yield();
        }
        this.requestCount++;
        return controllers_1.SendState.Send(() => {
            this.replyCount++;
            if (this.replyCount === 1) {
                return controllers_1.SendState.Yield();
            }
            const didConnect = this.establishmentController.didConnect();
            const assetDetails = this.assetController.getDestinationAsset();
            return !didConnect
                ? controllers_1.SendState.Error(__1.PaymentError.EstablishmentFailed)
                : !assetDetails
                    ? controllers_1.SendState.Error(__1.PaymentError.UnknownDestinationAsset)
                    : controllers_1.SendState.Done(assetDetails);
        });
    }
}
exports.AssetProbe = AssetProbe;
