// @interledger/pay/dist/src/senders/rate-probe.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RateProbe = void 0;
const __1 = require("@interledger/pay");
const controllers_1 = require("@interledger/pay/dist/src/controllers");
const utils_1 = require("@interledger/pay/dist/src/utils");
const max_packet_1 = require("@interledger/pay/dist/src/controllers/max-packet");
const exchange_rate_1 = require("@interledger/pay/dist/src/controllers/exchange-rate");
const sequence_1 = require("@interledger/pay/dist/src/controllers/sequence");
const establishment_1 = require("@interledger/pay/dist/src/controllers/establishment");
const expiry_1 = require("@interledger/pay/dist/src/controllers/expiry");
const failure_1 = require("@interledger/pay/dist/src/controllers/failure");
const asset_details_1 = require("@interledger/pay/dist/src/controllers/asset-details");
const pacer_1 = require("@interledger/pay/dist/src/controllers/pacer");
const _1 = require("@interledger/pay/dist/src/senders");
class RateProbe extends _1.StreamSender {
    constructor({ plugin, destination }) {
        super(plugin, destination);
        this.remainingTestAmounts = [
            utils_1.Int.ZERO,
            utils_1.Int.from(10 ** 12),
            utils_1.Int.from(10 ** 11),
            utils_1.Int.from(10 ** 10),
            utils_1.Int.from(10 ** 9),
            utils_1.Int.from(10 ** 8),
            utils_1.Int.from(10 ** 7),
            utils_1.Int.from(10 ** 6),
            utils_1.Int.from(10 ** 5),
            utils_1.Int.from(10 ** 4),
            utils_1.Int.from(10 ** 3),
        ];
        this.inFlightAmounts = new Set();
        const { requestCounter } = destination;
        this.rateCalculator = new exchange_rate_1.ExchangeRateController();
        this.maxPacketController = new max_packet_1.MaxPacketAmountController();
        this.controllers = [
            new sequence_1.SequenceController(requestCounter),
            new establishment_1.EstablishmentController(destination),
            new expiry_1.ExpiryController(),
            new failure_1.FailureController(),
            this.maxPacketController,
            new asset_details_1.AssetDetailsController(destination),
            new pacer_1.PacingController(),
            this.rateCalculator,
        ];
    }
    nextState(request) {
        if (!this.deadline) {
            this.deadline = Date.now() + RateProbe.TIMEOUT;
        }
        else if (Date.now() > this.deadline) {
            request.log.error('rate probe failed. did not establish rate and/or path capacity');
            return controllers_1.SendState.Error(__1.PaymentError.RateProbeFailed);
        }
        const probeAmount = this.remainingTestAmounts.shift();
        if (!probeAmount || this.inFlightAmounts.has(probeAmount.value)) {
            return controllers_1.SendState.Yield();
        }
        request.setSourceAmount(probeAmount);
        this.inFlightAmounts.add(probeAmount.value);
        return controllers_1.SendState.Send(() => {
            var _a;
            this.inFlightAmounts.delete(probeAmount.value);
            const nextProbeAmount = (_a = this.maxPacketController.getNextMaxPacketAmount()) !== null && _a !== void 0 ? _a : probeAmount;
            if (!this.remainingTestAmounts.some((n) => n.isEqualTo(nextProbeAmount))) {
                this.remainingTestAmounts.push(nextProbeAmount);
            }
            return this.maxPacketController.isProbeComplete()
                ? controllers_1.SendState.Done({
                    lowEstimatedExchangeRate: this.rateCalculator.getLowerBoundRate(),
                    highEstimatedExchangeRate: this.rateCalculator.getUpperBoundRate(),
                    maxPacketAmount: this.maxPacketController.getMaxPacketAmountLimit(),
                })
                : controllers_1.SendState.Schedule();
        });
    }
}
exports.RateProbe = RateProbe;
RateProbe.TIMEOUT = 10000;
RateProbe.MAX_PROBE_AMOUNT = utils_1.Int.from(1000000000000);
